OutOfSchool/OutOfSchool.Redis/CacheService.cs

```diff
@@ -170,4 +134,27 @@ private async Task ExecuteRedisMethod(Action redisMethod)
             }
         }
     }
+
+    private Task SetAsync<T>(
```

Comment:
(Performance) As I see the `SetAsync` is used in one place, thus can be inlined there with moving all code under one `ExecuteRedisMethod` lambda. The less separate lambdas, separate tasks, context switching - the faster it is.

As result, you could avoid excessive resetting cache value (when it is == default(T)) and rewrite that `if (EqualityComparer<T>.Default.Equals(returnValue, default))` part of logic in similar way as was suggested for `MultiLayerCache`.

Reply:
I could look like
```
    public async Task<T> GetOrAddAsync<T>(
        string key,
        Func<Task<T>> newValueFactory,
        TimeSpan? absoluteExpirationRelativeToNowInterval = null,
        TimeSpan? slidingExpirationInterval = null)
    {
        T returnValue = default;

        await ExecuteRedisMethod(async () =>
        {
            cacheLock.EnterUpgradeableReadLock();
            try
            {
                var value = await cache.GetStringAsync(key);

                if (value != null)
                {
                    returnValue = JsonConvert.DeserializeObject<T>(value);
                    return;
                }

                cacheLock.EnterWriteLock();
                try
                {
                    returnValue = await newValueFactory();
                    var options = new DistributedCacheEntryOptions
                    {
                        AbsoluteExpirationRelativeToNow = absoluteExpirationRelativeToNowInterval ?? redisConfig.AbsoluteExpirationRelativeToNowInterval,
                        SlidingExpiration = slidingExpirationInterval ?? redisConfig.SlidingExpirationInterval,
                    };

                    await cache.SetStringAsync(key, JsonConvert.SerializeObject(returnValue), options);
                }
                finally
                {
                    cacheLock.ExitWriteLock();
                }
            }
            finally
            {
                cacheLock.ExitUpgradeableReadLock();
            }
        });

        return returnValue;
    }
```

That could be further simplified having `Task<T> ExecuteRedisMethod<T>(Func<Task<T>>)` version implemented.

Reply:
Deleted set.